### PR TITLE
planner: check null and empty for `!= any(subq)` and `= all(subq)`

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/8798

### What is changed and how it works?

When converting "compare subquery" in expression rewriter, build additional 2 conditions to attach to projection:
- check if subquery is empty; if subquery is empty, `!= any(subq)` would always return false, while `= all(subq)` would always return true;
- check if outer key is null when subquery is not empty; if outer key is null, both `!= any(subq)` and `= all(subq)` would return null; note that it may be counterintuitive that `null = all(1, 2)` should return null, but it indeed holds and is verified on MySQL. 

### Check List 

Tests 

 - Unit test

Code changes

N/A

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -428,44 +428,52 @@ func (er *expressionRewriter) handleOtherComparableSubq(lexpr, rexpr expression.
...
-		funcCount := aggregation.NewAggFuncDesc(er.ctx, ast.AggFuncCount, []expression.Expression{funcIsNull}, false)
-		colCount := &expression.Column{
-			ColName:  model.NewCIStr("agg_col_cnt"),
-			UniqueID: er.ctx.GetSessionVars().AllocPlanColumnID(),
-			RetType:  funcCount.RetTp,
-		}
-		plan4Agg.AggFuncs = append(plan4Agg.AggFuncs, funcCount)
-		plan4Agg.schema.Append(colCount)
 		// All of the inner record set should not contain null value. So for t.id < all(select s.id from s), it
 		// should be rewrote to t.id < min(s.id) and if(sum(s.id is null) = 0, true, null).
```
Comment (winoros) [inline planner/core/expression_rewriter.go]: Update this comment?

File: planner/core/expression_rewriter.go [hunk 2]
```diff
@@ -428,44 +428,52 @@ func (er *expressionRewriter) handleOtherComparableSubq(lexpr, rexpr expression.
...
+		emptyChecker := expression.NewFunctionInternal(er.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), colCount, expression.Zero)
+		// If outer key is null, and subquery is not empty, it should always return null, even when it is `null = all (1, 2)`.
+		outerNullChecker := expression.NewFunctionInternal(er.ctx, ast.If, types.NewFieldType(mysql.TypeTiny), outerIsNull, expression.Null, expression.Zero)
+		cond = expression.ComposeDNFCondition(er.ctx, cond, emptyChecker, outerNullChecker)
 	} else {
-		// For "any" expression, if the record set has null and the cond return false, the result should be NULL.
-		hasNull := expression.NewFunctionInternal(er.ctx, ast.NE, types.NewFieldType(mysql.TypeTiny), colSum, expression.Zero)
-		nullChecker := expression.NewFunctionInternal(er.ctx, ast.If, types.NewFieldType(mysql.TypeTiny), hasNull, expression.Null, expression.Zero)
-		cond = expression.ComposeDNFCondition(er.ctx, cond, nullChecker)
+		// For "any" expression, if the subquery has null and the cond return false, the result should be NULL.
```
Comment (winoros) [inline planner/core/expression_rewriter.go]: Add a rewrite result like line 457?

File: planner/core/expression_rewriter.go [hunk 3]
```diff
@@ -429,44 +429,53 @@ func (er *expressionRewriter) handleOtherComparableSubq(lexpr, rexpr expression.
...
 	colSum := &expression.Column{
 		ColName:  model.NewCIStr("agg_col_sum"),
 		UniqueID: er.ctx.GetSessionVars().AllocPlanColumnID(),
 		RetType:  funcSum.RetTp,
 	}
 	plan4Agg.AggFuncs = append(plan4Agg.AggFuncs, funcSum)
 	plan4Agg.schema.Append(colSum)
+	innerHasNull := expression.NewFunctionInternal(er.ctx, ast.NE, types.NewFieldType(mysql.TypeTiny), colSum, expression.Zero)
+
+	funcCount := aggregation.NewAggFuncDesc(er.ctx, ast.AggFuncCount, []expression.Expression{innerIsNull}, false)
```
Comment (lamxTyler) [inline planner/core/expression_rewriter.go]: Is it `count(1)`?

Comment (eurekaka) [reply planner/core/expression_rewriter.go]: Yes, it is used to check if the inner plan is empty.

Comment (eurekaka) [reply planner/core/expression_rewriter.go]: Done, PTAL

**Pull request merged**